### ISSUE TITLE
configuration file detection is broken on Windows

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#661]
In GitLab by @fkohlgrueber on Dec 5, 2018, 11:12

### Please describe how you installed Flake8

- Windows 10
- Python 3.7.1

```
C:\Users\felix\git\flake8_bug>python -m pip install flake8
...
C:\Users\felix\git\flake8_bug>python -m flake8 --version
3.6.0 (mccabe: 0.6.1, pycodestyle: 2.4.0, pyflakes: 2.0.0) CPython 3.7.1 on Windows
C:\Users\felix\git\flake8_bug>python -m flake8 --bug-report
{
  "dependencies": [
    {
      "dependency": "setuptools",
      "version": "40.6.2"
    }
  ],
  "platform": {
    "python_implementation": "CPython",
    "python_version": "3.7.1",
    "system": "Windows"
  },
  "plugins": [
    {
      "is_local": false,
      "plugin": "mccabe",
      "version": "0.6.1"
    },
    {
      "is_local": false,
      "plugin": "pycodestyle",
      "version": "2.4.0"
    },
    {
      "is_local": false,
      "plugin": "pyflakes",
      "version": "2.0.0"
    }
  ],
  "version": "3.6.0"
}
```

### Please describe the problem

The discovery of configuration files (e.g. `tox.ini`) doesn't work on Windows in a couple of cases. This can be demonstrated in the following minimal example project:

```
flake8_bug (project root folder)
│   tox.ini
└───testing
    │   file_a.py
    │   file_b.py
    └───subfolder
            file_c.py
```

**Content of file_(a/b/c).py**

```
"""
This is a very long line that exceeds the standard line length limit of 79 characters.
"""

```

**Content of tox.ini**

```
[flake8]
max-line-length = 120
```

### Expected outcome

Since the line length is expanded to 120 characters in the configuration file, there shouldn't be any errors when running flake8.

On Ubuntu, this works as expected (no errors):

```
felix@the-machine:/mnt/c/Users/felix/git/flake8_bug$ python3 -m flake8 --version
3.6.0 (mccabe: 0.6.1, pycodestyle: 2.4.0, pyflakes: 2.0.0) CPython 3.6.7 on Linux
felix@the-machine:/mnt/c/Users/felix/git/flake8_bug$ python3 -m flake8
felix@the-machine:/mnt/c/Users/felix/git/flake8_bug$ python3 -m flake8 testing
felix@the-machine:/mnt/c/Users/felix/git/flake8_bug$ python3 -m flake8 testing/
felix@the-machine:/mnt/c/Users/felix/git/flake8_bug$ python3 -m flake8 testing/file_a.py
felix@the-machine:/mnt/c/Users/felix/git/flake8_bug$ python3 -m flake8 testing/file_a.py testing/file_b.py
felix@the-machine:/mnt/c/Users/felix/git/flake8_bug$ python3 -m flake8 testing/file_a.py testing/file_b.py testing/subfolder/file_c.py
felix@the-machine:/mnt/c/Users/felix/git/flake8_bug$ python3 -m flake8 testing/file_a.py testing/subfolder/file_c.py
felix@the-machine:/mnt/c/Users/felix/git/flake8_bug$
```

### Actual outcome on Windows

If I do the same on Windows, the configuration file isn't used in some cases (which leads to "line too long" errors):

```
C:\Users\felix\git\flake8_bug>python -m flake8

C:\Users\felix\git\flake8_bug>python -m flake8 testing

C:\Users\felix\git\flake8_bug>python -m flake8 testing\
testing\file_a.py:2:80: E501 line too long (86 > 79 characters)
testing\file_b.py:2:80: E501 line too long (86 > 79 characters)
testing\subfolder\file_c.py:2:80: E501 line too long (86 > 79 characters)

C:\Users\felix\git\flake8_bug>python -m flake8 testing\file_a.py

C:\Users\felix\git\flake8_bug>python -m flake8 testing\file_a.py testing\file_b.py

C:\Users\felix\git\flake8_bug>python -m flake8 testing\file_a.py testing\file_b.py testing\subfolder\file_c.py
testing\file_a.py:2:80: E501 line too long (86 > 79 characters)
testing\file_b.py:2:80: E501 line too long (86 > 79 characters)
testing\subfolder\file_c.py:2:80: E501 line too long (86 > 79 characters)

C:\Users\felix\git\flake8_bug>python -m flake8 testing\file_a.py testing\subfolder\file_c.py
testing\file_a.py:2:80: E501 line too long (86 > 79 characters)
testing\subfolder\file_c.py:2:80: E501 line too long (86 > 79 characters)

C:\Users\felix\git\flake8_bug>
```

### Possible cause

The problem seems to be in the `ConfigFileFinder`, especially in its constructor and the `generate_possible_local_files` method (`options/config.py:96`).

In line 55, `self.parent` and `self.tail` are initialized with `os.path.abspath(os.path.commonprefix(args))`, which works differently on Windows / Unix. On Windows, the path contains a trailing backslash. When `os.split` is called on line 110, the result is also different:

```
Windows:
Python 3.7.1 (v3.7.1:260ec2c36a, Oct 20 2018, 14:57:15) [MSC v.1915 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.path.abspath(os.path.commonprefix([r"testing\file_a.py", r"testing\subfolder\file_c.py"]))
'C:\\Users\\felix\\git\\flake8_bug\\testing\\'
>>> os.path.split(_)
('C:\\Users\\felix\\git\\flake8_bug\\testing', '')
>>>

Ubuntu:
Python 3.6.7 (default, Oct 22 2018, 11:32:17)
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.path.abspath(os.path.commonprefix(["testing/file_a.py", "testing/subfolder/file_c.py"]))
'/mnt/c/Users/felix/git/flake8_bug/testing'
>>> os.path.split(_)
('/mnt/c/Users/felix/git/flake8_bug', 'testing')
>>>
```

The empty `tail` that Windows produces ends the upwards search for configuration files. There might be other issues that cause Windows to behave differently than Unix.

Is this a known issue? I'd be interested in working on a fix for this but wanted to hear your input first.

Thanks in advance.

**Comment by asottile** [comment]
In GitLab by @asottile on Dec 5, 2018, 11:19

your analysis looks correct -- does an `.lstrip(os.sep)` fix the problem (after the `abspath`)?

**Comment by asottile** [comment]
In GitLab by @fkohlgrueber on Dec 5, 2018, 11:33

Solving the issue might be easier than I thought. Simply replacing `os.path.commonprefix` with `os.path.commonpath` solves the issue in the examples I posted.

IMO, wsing `os.path.commonprefix` doesn't make sense because it even yields prefixes for non-existing folders:

```
>>> os.path.commonprefix(["testing/subfolder/file_c.py", "test_folder/file_a.py"])
'test'
>>> os.path.commonpath(["testing/subfolder/file_c.py", "test_folder/file_a.py"])
''
```

Using `os.path.commonpath` yields a path without trailing separators on both Windows and Unix:

```
Windows:
>>> os.path.commonpath([r"testing\subfolder\file_c.py", r"testing\file_a.py"])
'testing'
>>> os.path.abspath(_)
'C:\\Users\\felix\\git\\flake8_bug\\testing'

Ubuntu:
>>> os.path.commonpath(["testing/subfolder/file_c.py", "testing/file_a.py"])
'testing'
>>> os.path.abspath(_)
'/mnt/c/Users/felix/git/flake8_bug/testing'
```

`.rstrip(os.sep)` seems to solve the issue as well, but I think that `os.path.commonpath` might be the better way to go. What do you think?

**Comment by asottile** [comment]
In GitLab by @asottile on Dec 5, 2018, 11:38

Wow that sounds even better! I didn't know about commonpath!

**Comment by asottile** [comment]
In GitLab by @fkohlgrueber on Dec 5, 2018, 11:40

Yeah, I found it by accident ;-)

Do you want me to open a PR with the change?

**Comment by asottile** [comment]
In GitLab by @asottile on Dec 5, 2018, 11:44

yeah that would be great!  If possible, could you also write a small test -- it sounds to me like your second example with `commonprefix` will trigger an issue even on posix (`os.path.commonprefix(["testing/subfolder/file_c.py", "test_folder/file_a.py"])`).  If not I can probably help you out with the test :)

**Comment by asottile** [comment]
In GitLab by @fkohlgrueber on Dec 5, 2018, 12:25

I did some more research and found that the trailing backslash on Windows is a bug that's going to be fixed in the next minor versions of Python 3.6 / 3.7, see [here](https://bugs.python.org/issue31047).

This bug also breaks some tests of flake8 on the latest python 3.6.7 and 3.7.1 on Windows. Since appveyor only tests up to python 3.5 currently, this has most likely not been recognized...

Also, `os.path.commonpath` is 3.5+ unfortunately.

**Comment by asottile** [comment]
In GitLab by @asottile on Dec 5, 2018, 12:27

oh interesting, so maybe not worth fixing if it's just affecting 3.6.7 and 3.7.1 and will be fixed in 3.6.8 / 3.7.2?

**Comment by asottile** [comment]
In GitLab by @fkohlgrueber on Dec 5, 2018, 12:49

Yep. And I think changing the appveyor config to test python 3.6 and 3.7 would be a good idea.

**Comment by asottile** [comment]
In GitLab by @asottile on Jan 3, 2019, 17:04

3.7.2 is out so this should be fixed -- thanks for the issue!

**Comment by asottile** [comment]
In GitLab by @asottile on Jan 3, 2019, 17:04

closed

**Comment by asottile** [comment]
In GitLab by @fkohlgrueber on Jan 3, 2019, 23:27

Yep, I just confirmed that the issue is fixed in 3.7.2.
